go/consensus/tendermint/apps/beapoch/backend_scrape.go

```diff
@@ -0,0 +1,431 @@
+	params *beapochState.ConsensusParameters,
+	scrapeState *beapochState.SCRAPEState,
+	epoch epochtime.EpochTime,
+) error {
+	if scrapeState != nil {
+		// XXX: Disqualify all the asshats that failed in the previous round.
+	}
+
+	if scrapeState == nil || scrapeState.Epoch != epoch {
+		newState := &beapochState.SCRAPEState{
+			Epoch: epoch,
+		}
+		scrapeState = newState
+	} else {
+		// The previous attempt to generate a beacon for this epoch failed.
+		scrapeState.Round++
+	}
+
+	// Draw participants.
+	schedulerState := schedulerState.NewMutableState(ctx.State())
+	registryState := registryState.NewMutableState(ctx.State())
+	validators, err := schedulerState.CurrentValidators(ctx)
+	if err != nil {
+		return fmt.Errorf("beapoch: failed to get current validators: %w", err)
+	}
+	candidateParticipants, err := validatorsByVotingPower(validators)
+	if err != nil {
+		return fmt.Errorf("beapoch: failed to sort current validators: %w", err)
+	}
+
+	const tempNumParticipants = 10
+	participants := make([]scrape.Point, 0, tempNumParticipants)
+	for _, validatorID := range candidateParticipants {
+		var node *node.Node
+		node, err = registryState.NodeByConsensusAddress(ctx, validatorID[:])
+		if err != nil || node.Beacon == nil {
+			continue
+		}
+
+		participants = append(participants, node.Beacon.Point)
+		if len(participants) == tempNumParticipants {
+			break
+		}
+	}
+	if l := len(participants); l < tempNumParticipants {
+		return fmt.Errorf("beapoch: insufficient beacon participants: %d (want %d)", l, tempNumParticipants)
+	}
+
+	// Initialize the SCRAPE state.
+	const tempThreshold = 6
+	if scrapeState.Instance, err = scrape.New(&scrape.Config{
+		Participants:  participants,
+		Threshold:     tempThreshold,
+		PVSSThreshold: tempThreshold,
+	}); err != nil {
+		return fmt.Errorf("beapoch: failed to initialize SCRAPE instance: %w", err)
+	}
+
+	// Derive the deadlines.
+	const (
+		tempCommitInterval  = 100
+		tempRevealInterval  = 90
+		tempTransitionDelay = 10
+	)
+	height := ctx.BlockHeight() + 1 // Current height is ctx.BlockHeight() + 1
+	scrapeState.CommitDeadline = height + tempCommitInterval
+	scrapeState.RevealDeadline = height + tempRevealInterval
+	scrapeState.TransitionHeight = height + tempCommitInterval + tempRevealInterval + tempTransitionDelay
+
+	// TODO: Broadcast an event to trigger commits.
+
+	// Store the round state.
+	if err := state.SetSCRAPEState(ctx, scrapeState); err != nil {
+		return fmt.Errorf("beapoch: failed to set SCRAPE state: %w", err)
+	}
+
+	ctx.Logger().Info("initializing SCRAPE round",
+		"epoch", epoch,
+		"round", scrapeState.Round,
+	)
+
+	return nil
+}
+
+func (impl *backendSCRAPE) ExecuteTx(
+	ctx *api.Context,
+	state *beapochState.MutableState,
+	params *beapochState.ConsensusParameters,
+	tx *transaction.Transaction,
+) error {
+	if tx.Method != MethodSCRAPEReveal && tx.Method != MethodSCRAPECommit {
+		return fmt.Errorf("beapoch: invalid method: %s", tx.Method)
+	}
+
```

Comment:
Gas costs? (I know that in practice only validators can call these methods, so they could opt to pay zero gas by only including these transactions when they are the proposer.)